# **Introduction for the assignment**





# **Part 1. Explanatory data analysis**

In this part we will have a quick overview of our data set BADS_WS2021_known.csv



---
**A. Dataset importation**


In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os #operating system interface

In [5]:
from google.colab import files
data_to_load = files.upload()

Saving BADS_WS2021_known.csv to BADS_WS2021_known (1).csv


In [6]:
import io
df = pd.read_csv(io.BytesIO(data_to_load['BADS_WS2021_known.csv']))
df #we check if the table is rightly imported

order_item_id  order_date  ... user_reg_date  return
0                  1  2016-06-22  ...    2016-06-23       0
1                  2  2016-06-22  ...    2016-06-23       0
2                  3  2016-06-22  ...    2015-03-15       1
3                  4  2016-06-22  ...    2015-03-15       0
4                  5  2016-06-22  ...    2015-03-15       1
...              ...         ...  ...           ...     ...
99995          99996  2016-09-11  ...    2015-02-17       0
99996          99997  2016-09-11  ...    2015-02-17       0
99997          99998  2016-09-11  ...    2016-09-10       0
99998          99999  2016-09-11  ...    2016-09-10       0
99999         100000  2016-09-11  ...    2015-02-17       0

[100000 rows x 14 columns]

**B. Dataset eyeballing**

In [7]:
print("Dimension of the data set is {}".format(df.shape))

Dimension of the data set is (100000, 14)


With function shape we see that our data set is of 14 columns and 100.000 rows. In terms of data frame, that represents 14 features and 100.000 observations.

In [8]:
df.head()

order_item_id  order_date  ... user_reg_date  return
0              1  2016-06-22  ...    2016-06-23       0
1              2  2016-06-22  ...    2016-06-23       0
2              3  2016-06-22  ...    2015-03-15       1
3              4  2016-06-22  ...    2015-03-15       0
4              5  2016-06-22  ...    2015-03-15       1

[5 rows x 14 columns]

Function head displays the first 5 rows so we can have a quick overlook on how the table is constructed. Let's analyse that.

1. Order_item_id : begins at 1 and add +1 with every new row, it is simply the count of purchases.

2. Order_date: date format YYYY-MM-DD of the order.

3. Delivery_date: date format YYYY-MM-DD of the delivery.

4. Item-id: 3 random digits that are supposedly representing one specific item each time.

5. Item_size: composed of numbers and characters, size ordered.

6. Item_color: characters, color ordered.

7. Brand_id: the id of the brand expressed in number (code).

8. Item_price: a continuous number, indicates the price of the item.

9. User_id: 5 digits that indicates the user's id.

10. User_title: Characters, indicates the user's title (Mrs, Mr...).

11. User_dob: format YYYY-MM-DD, indicates date of birth of the customer.

12. User_state: characters that display living state of the customer.

13. User_reg_date: format YYYY-MM-DD, it is the registration date, let us know customer's seniority. 

14. Return: binary types, 1 if article is returned, 0 if not.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 14 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   order_item_id  100000 non-null  int64  
 1   order_date     100000 non-null  object 
 2   delivery_date  90682 non-null   object 
 3   item_id        100000 non-null  int64  
 4   item_size      100000 non-null  object 
 5   item_color     100000 non-null  object 
 6   brand_id       100000 non-null  int64  
 7   item_price     100000 non-null  float64
 8   user_id        100000 non-null  int64  
 9   user_title     100000 non-null  object 
 10  user_dob       91275 non-null   object 
 11  user_state     100000 non-null  object 
 12  user_reg_date  100000 non-null  object 
 13  return         100000 non-null  int64  
dtypes: float64(1), int64(5), object(8)
memory usage: 10.7+ MB


We observe that our data set is made of multiple types:

- intergers

- floats

- objects

The memory usage is of 10MB, we'll see if we can improve the performance later.
Moreover, we observe missing values for the features delivery_date and user_dob.

In [10]:
df.describe()

order_item_id        item_id  ...        user_id         return
count  100000.000000  100000.000000  ...  100000.000000  100000.000000
mean    50000.500000     923.781820  ...   27169.073960       0.458180
std     28867.657797     702.568822  ...   14053.417669       0.498251
min         1.000000       1.000000  ...       9.000000       0.000000
25%     25000.750000     214.000000  ...   14937.000000       0.000000
50%     50000.500000     812.000000  ...   31189.000000       0.000000
75%     75000.250000    1581.000000  ...   38917.250000       1.000000
max    100000.000000    2241.000000  ...   48241.000000       1.000000

[8 rows x 6 columns]

The function describe gives us statistique insight of our data set for each characteristic that is numerical. Some of the result are not relevant.

E.g. Order_item_id is perfectly distributed as it is the count of purchases.

E.g.2. Return being a binary type does not give distribution information.

Some features are more talkative. For instance, Item_price displays a mean larger than the median. We also observe that the max value is 999 while the 3rd quartile is only 79.9. Therefore we expect to find significant outliers.

In [11]:
df.nunique(axis=0)

order_item_id    100000
order_date           82
delivery_date       123
item_id            1943
item_size           102
item_color           77
brand_id            135
item_price          252
user_id           19205
user_title            5
user_dob           9247
user_state           16
user_reg_date       573
return                2
dtype: int64

The function unique let us know how many different variables we'll find in each feature. We know now that 135 different brands are sold by the shop, 19205 customers from 16 states have ordered 100.000 items, etc.

**C. Target eyeballing**

Now we want to know more about our target variable. First action here is to change the column's name return to returns. This will avoid miscomprehension from Python with the function return

In [12]:
df.rename(columns = {'return': 'returns'})
df = df.rename(columns = {'return': 'returns'})

We can now have a better look over our Target variable, returns.



In [13]:
df.returns.value_counts()

0    54182
1    45818
Name: returns, dtype: int64

We see that almost half of the items are return, which is a pretty tragic amount. We can aggregate this information with other variables. Let's find the mode in price, item and brand.

In [14]:
mode_price = df.groupby("returns")["item_price"].apply(lambda x: x.mode().iloc[0])
mode_item = df.groupby("returns")["item_id"].apply(lambda x: x.mode().iloc[0])
mode_brand = df.groupby("returns")["brand_id"].apply(lambda x: x.mode().iloc[0])
print(mode_price)
print(mode_item)
print(mode_brand)

returns
0    39.9
1    59.9
Name: item_price, dtype: float64
returns
0    1415
1     100
Name: item_id, dtype: int64
returns
0    3
1    3
Name: brand_id, dtype: int64


Here we see that the most returned item price is 59.9€, the item the most returned is the number 100 and the brand that is the most returned is the 3. We can observed that brand 3 is both the most returned and the most kept, we could imagine that being the most sold brand, the chances of returned are simply higher.

In [15]:
df.groupby("returns")["brand_id"].value_counts()

returns  brand_id
0        3           7214
         1           4279
         5           3066
         37          2683
         11          1800
                     ... 
1        100            1
         105            1
         136            1
         137            1
         138            1
Name: brand_id, Length: 254, dtype: int64

As expected, counts of brand_id shows us that 3 is the most sold brand by far.

# **Part 2. Data preparation**

Now that we know what we've put our foot in, we can start preparing our dataset for prediction models. 

**A. Data types**

We're gonna start our preparation by altering some of the data types so that our data set is more efficient.

First thing we do is changing all our Int64 and float64 to float32.

In [16]:
#We start by changing float64 in float32
num_vars = df.select_dtypes(include=np.float64).columns
df[num_vars] = df[num_vars].astype(np.float32)

#We change Int64 in float32
num_vars = df.select_dtypes(include=np.int64).columns
df[num_vars] = df[num_vars].astype(np.float32)

Because we know that returns is a binary variable we're gonna store it as boolean.

In [17]:
df["returns"] = df["returns"].astype("bool")

Now we're focusing on user_title that only contains 5 variables.

In [18]:
df.user_title.value_counts()

Mrs             95429
Mr               3915
Family            414
Company           128
not reported      114
Name: user_title, dtype: int64

Because Family, Company and Not reported are so small, I choose to group them as "Else" and to transform user_title as a 3 multinomial variable.

In [19]:
df = df.replace("Family", "Else")
df = df.replace("Company", "Else")
df = df.replace("not reported", "Else")
print(df.user_title.value_counts())

df["user_title"] = df["user_title"].astype("category")

Mrs     95429
Mr       3915
Else      656
Name: user_title, dtype: int64


Within our features, 4 of them are format date YYYY-MM-DD. We want to change them in date type.



In [20]:
df["order_date"] = pd.to_datetime(df['order_date'])
df["delivery_date"] = pd.to_datetime(df['delivery_date'])
df["user_dob"] = pd.to_datetime(df['user_dob'])
df["user_reg_date"] = pd.to_datetime(df['user_reg_date'])

We transform the 3 remaining features (state, size and color) in categories.

In [21]:
df["user_state"] = df["user_state"].astype("category")
df["item_size"] = df["item_size"].astype("category")
df["item_color"] = df["item_color"].astype("category")

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 14 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   order_item_id  100000 non-null  float32       
 1   order_date     100000 non-null  datetime64[ns]
 2   delivery_date  90682 non-null   datetime64[ns]
 3   item_id        100000 non-null  float32       
 4   item_size      100000 non-null  category      
 5   item_color     100000 non-null  category      
 6   brand_id       100000 non-null  float32       
 7   item_price     100000 non-null  float32       
 8   user_id        100000 non-null  float32       
 9   user_title     100000 non-null  category      
 10  user_dob       91275 non-null   datetime64[ns]
 11  user_state     100000 non-null  category      
 12  user_reg_date  100000 non-null  datetime64[ns]
 13  returns        100000 non-null  bool          
dtypes: bool(1), category(4), datetime64[ns](4), float32(5

We see here that all data types have been righteously transformed. In the process we've save almost half of our memory usage.

**B. Dataset improvement**

Because it might end up worhty, or maybe not I'ved decided to insert two new features to our dataset:

- How many time the delivary has taken: delivery_time
- How old are the customers: user_age

In [ ]:
delivery_time = df["delivery_date"] - df["order_date"]
df.add(delivery_time)
df["delivery_time"] = pd.to_timedelta(df['delivery_time'])